src/coreclr/inc/corinfo.h

```diff
@@ -3194,7 +3194,8 @@ class ICorDynamicInfo : public ICorStaticInfo
     // return the data's address (for static fields only)
     virtual void* getFieldAddress(
                     CORINFO_FIELD_HANDLE    field,
-                    void                  **ppIndirection = NULL
+                    void                  **ppIndirection = NULL,
+                    void                  **pBoxedStatic = NULL
```

Comment:
Returning pointer has ambiguity. For example, what does it mean when `*ppIndirection` is non-null and `*ppBoxedStatic` is non-null?

I think returning `bool` that says whether the pointer needs unboxing would be better.

Reply:
@jkotas The problem that currently in importer we call `getFieldInfo` and for structs we assume that we'll need unboxing so we end up with a tree like this (we don't know in advance that we won't need unboxing):
![image](https://user-images.githubusercontent.com/523221/200025358-6e46ea08-70a8-4b61-95f0-f4350db0032a.png)

From my understanding, we might potentially even remove it before morph in case of dead code so we don't have to always call `getFieldAddress` for it (hence, the populate class stuff). Then in morph we transform `FIELD` into either `IND(handle-to-boxed-static)` or `CNS_INT` which is a direct handle to the boxed static. Since we don't use `TYP_REF` here, that ADD(CNS_INT, 8) is then folded.

`bool` argument slightly complicates this design as when we're inspecting `FIELD` in morph we can't access its parent to remove. We can visit GT_ADD node but who guarantees it wasn't changed to something else? If you still insist on a bool arg I might try to rework it.

Reply:
It might work easily if we can afford `getFieldAddress` early in importer to figure out whether it's unboxed already. But that leads to the question on why we can't call populate class directly in getFieldInfo then. Maybe we can get back to the `getFieldInfo` and limit to "initialized types"?

Reply:
NOTE: the current PR should work carefully with non-initialized types as far as I understand, the flag `GTF_ICON_INITCLASS` should tell jit to avoid using the load before initialization